helix-vcs/src/lib.rs

```diff
@@ -21,129 +16,186 @@ mod status;
-        self.providers
-            .iter()
-            .find_map(|provider| match provider.get_current_head_name(file) {
-                Ok(res) => Some(res),
-                Err(err) => {
-                    log::debug!("{err:#?}");
-                    log::debug!("failed to obtain current head name for {}", file.display());
-                    None
-                }
-            })
+    pub fn get_current_head_name(&self, file: &Path) -> Option<DiffHead> {
+        debug_assert!(!file.exists() || file.is_file());
+        debug_assert!(file.is_absolute());
+
+        let res = match self {
+            Self::None => return None,
+            Self::File => return None,
+            #[cfg(feature = "git")]
+            Self::Git => git::get_current_head_name(file),
+        };
+
+        match res {
+            Ok(o) => Some(o),
+            Err(e) => {
+                log::debug!("failed to current head name for {}: {e:#?}", file.display());
+                None
+            }
+        }
     }
 
-    /// Fire-and-forget changed file iteration. Runs everything in a background task. Keeps
-    /// iteration until `on_change` returns `false`.
+    /// Fire-and-forget changed file iteration. Runs everything in a background task.
+    /// Keeps iterating until `on_change` returns `false`.
     pub fn for_each_changed_file(
         self,
         cwd: PathBuf,
-        f: impl Fn(Result<FileChange>) -> bool + Send + 'static,
+        on_change: impl Fn(Result<FileChange>) -> bool + Send + 'static,
     ) {
         tokio::task::spawn_blocking(move || {
-            if self
-                .providers
-                .iter()
-                .find_map(|provider| provider.for_each_changed_file(&cwd, &f).ok())
-                .is_none()
-            {
-                f(Err(anyhow!("no diff provider returns success")));
+            let res = match self {
+                Self::None => Ok(()),
+                Self::File => Ok(()),
+                #[cfg(feature = "git")]
+                Self::Git => git::for_each_changed_file(&cwd, &on_change),
+            };
+
+            if res.is_err() {
+                on_change(Err(anyhow!("no diff provider returned success")));
             }
         });
     }
-}
 
-impl Default for DiffProviderRegistry {
-    fn default() -> Self {
-        // currently only git is supported
-        // TODO make this configurable when more providers are added
-        let providers = vec![Git.into()];
-        DiffProviderRegistry { providers }
+    /// Used to offer completion options for the `diff-source` command
+    pub fn all_values() -> &'static [&'static str] {
+        // To force a miscompilation and remember to add the value to the list
+        match Self::None {
+            Self::None => (),
+            #[cfg(feature = "git")]
+            Self::File => (),
+            Self::Git => (),
+        }
+
+        // Keep it sorted alphabetically
+        &[
```

Comment:
I think splitting this into two Methods where one returns a list of all variants and one turns a variant into a string would be more generally useful (I will need the latter to turn this into a config option for example and you can reuse it in the display implementation)

I really wish rust had a builtin way to iterate all vsriantes of a fieldless enum. In my own projects I have a macro for that but it doesn't work with cfg and it's a bit of a pain to use since you need to put the entoe enum kn the macro... Or maybe one day we will have macro_rules based derive macros